### PR TITLE
Add configuration option for disabling automatic eager loading

### LINKED ISSUE #55: Add configuration for
Is there the ability to set goldiloader to set the whole application to `auto_include(false)` by default? I only want to use goldiloader for certain controller actions. I think I need something like:

```ruby
Goldiloader.config do |config|
 config.auto_include = false
end
```

Then when I want to use it I can do the following

```ruby
@posts = Post.auto_include # preferable
# or
@posts = Post.auto_include(true)
```

### PR DESCRIPTION
**Pull request by fsateler**
Fixes #55

This patch enables the auto loading via 

```ruby
Goldiloader.auto_include do
  # here, goldiloader is active
end
```

Based on the work of @gzunigal

### CODE REVIEW

File: lib/goldiloader/active_record_patches.rb [hunk 1]
```diff
@@ -47,7 +47,7 @@ def auto_include!(auto_include = true)
     end
 
     def auto_include_value
-      @values.fetch(:auto_include, true)
+      @values.fetch(:auto_include) { Goldiloader.auto_include }
```
Comment (jturkel) [inline lib/goldiloader/active_record_patches.rb:50]: 🤔  Just to make sure I understand the intent here: `auto_include` can be set at three different levels with the following order of precedence:

1. `ActiveRecord::Relation` instances either via regular ActiveRecord query scope or an association's scope
2. Within a thread
3. Globally (this wasn't actually implemented but I assume that was the intent based on the README change)

I believe the main scenarios that 2 and 3 are attempting to support (as described in #55) are:

* Goldiloader enabled by default but disabled for a block of code
* Goldiloader disabled by default but enabled for a block of code

This seems like it works reasonably well if `ActiveRecord::Relation` instances only explicitly set `auto_include = false` when automatic eager loading is known to be problematic. However, I'm concerned it could lead to unexpected results if `ActiveRecord::Relation` instances set `auto_include = true` since the developers intent as to how this code should behave in those main scenarios is less clear.

All of this makes me wonder if this solution is too complex. Perhaps a simpler alternative would be to just support a global configuration option which is the default value for `ActiveRecord::Relation#auto_include`? This would allow developers to incrementally rollout automatic eager loading association by association rather than code block by code block which is likely a better adoption strategy.

**Review by jturkel — CHANGES_REQUESTED** [review]
I'm still very much open to do something in this area but I'm concerned the current proposal is going to result in unexpected behavior i.e. bugs in goldiloader clients. I'm curious to hear your thoughts on a potential simplification and I'd love to really better understand the problem we're trying to solve so we can come up with the best solution.

File: README.md [hunk 2]
```diff
@@ -90,6 +92,20 @@ class Blog < ActiveRecord::Base
...
+You can also disable automatic eager loading completely through configuration:
+
+```ruby
+Goldiloader.auto_include = false
+```
+
+Then you can use it selectively with a block:
+
+```ruby
+Goldiloader.auto_include do
```
[the remark this answers is not in the document]
Comment (fsateler) [reply README.md:104]: This is by design (the idea is to keep server threads isolated), but there is a problem here: during boot, there is a single thread, and this change needs to be propagated there.

File: lib/goldiloader/active_record_patches.rb (hunk 1 above)
Comment (fsateler) [reply lib/goldiloader/active_record_patches.rb:50]: Hi! Before going through with more changes, I think it's best to get everyone aligned on what to do. This seems the best place to do that.

> Just to make sure I understand the intent here:

Personally, I only care about (dynamic) global disable/enable. This needs to be thread-aware, otherwise multi-threaded servers like puma would have the setting enabled/disabled by other threads.

Why do I only care about global, dynamic enabled/disable? Basically, I want to have it disabled by default, and enable it per request/per job when appropriate. The reasons for opt-in instead of opt-out are that some object graphs we deal with can cause extensive memory when preloading everything. Thus, we enable this on select requests/jobs (which is also how I read the intent expressed on https://github.com/salsify/goldiloader/issues/55#issue-230046390).

This PR (attempts to) preserve the `auto_include` handling, as requested by https://github.com/salsify/goldiloader/issues/55#issuecomment-427937927 .

I explicitly did not add "default enabled, disable per block" because I don't really find that scenario particularly useful, and keeps both the code and the API simpler.

Does this make sense? And does it make the code make more sense?

File: README.md [hunk 3]
```diff
@@ -90,6 +92,20 @@ class Blog < ActiveRecord::Base
 end
 ```
 
+You can also disable automatic eager loading completely through configuration:
+
+```ruby
+Goldiloader.auto_include = false
```
[the remark this answers is not in the document]
Comment (fsateler) [reply README.md]: Fixed via a `global_auto_include` method.

File: lib/goldiloader/active_record_patches.rb (hunk 1 above)
Comment (fsateler) [reply lib/goldiloader/active_record_patches.rb:50]: Hi! Friendly ping. I have updated to fix the threading issue.

> However, I'm concerned it could lead to unexpected results if ActiveRecord::Relation instances set auto_include = true since the developers intent as to how this code should behave in those main scenarios is less clear.

I agree this is not very clear, which is why we sidestepped the issue :)

Honestly, per-`ActiveRecord::Relation` is confusing to me: doing `objects.auto_include(false).to_a.first.relation_1.relation_2` would still eager load `relation_2`, which is very confusing. This is why I prefer a (threaded) block enable/disable version

Comment (jturkel) [reply lib/goldiloader/active_record_patches.rb:50]: Apologies for the delay in responding. I've been out on vacation and I'm digging out from a mountain of email. I'll give this some more thought and get back to you in a few days.

Comment (jturkel) [reply lib/goldiloader/active_record_patches.rb:50]: Thanks for your patience on this @fsateler. Thinking it through a bit more, a big source of my concern stems from reusing the `auto_include` term and how that interacts with `ActiveRecord::Relation#auto_include`. What do you think about something like this which will hopefully make this interaction more obvious?

```ruby
# Enable/disable Goldiloader globally across all threads
Goldiloader.enabled = false

# Enable Goldiloader for the current thread + block of code
Goldiloader.enabled do
  # ...
end

# Disable Goldiloader for the current thread + block of code
Goldiloader.disabled do
  # ...
end
```

Comment (fsateler) [reply lib/goldiloader/active_record_patches.rb:50]: Sounds good to me! I'll work on that.

Comment (fsateler) [reply lib/goldiloader/active_record_patches.rb:50]: Follow up questions:

1. I'm thinking the global switch should be called differently from the block switch. This is becase it should be possible to use non-block versions when you don't have a block (but still want to respect threads). For example, rspec before/after or minitest setup/teardown. What do you think?
2. What do you think goldiloader should do with relations that "escape" the enabled block?  That is: `Goldiloader.enabled{ $blogs = Blog.all } ; blogs.map(&:author)`. Should `author` be auto-included?

Comment (jturkel) [reply lib/goldiloader/active_record_patches.rb]: 1. Good point. A non-block form that only impacts the current thread would be useful (although for tests wouldn't you want to toggle the global setting so it impacts all threads?). The difference in behavior between `Goldiloader.disable!` and `Goldiloader.enable = val` is a bit subtle. I'd rather go with something that's more obvious even if it means a more verbose name. How about `Goldiloader.enabled=(val)` for the global version and then `Goldiloader.enabled_for_current_thread=(val)` with appropriate block forms for disabling/enabling in the current thread?

2. I agree that automatic eager loading should only happen if Goldiloader is enabled at the time that the association is actually traversed i.e. in the `Goldiloader.enabled{ $blogs = Blog.all } ; blogs.map(&:author)` example, `author` should not be automatically eager loaded.

Comment (fsateler) [reply lib/goldiloader/active_record_patches.rb]: Thanks! I have pushed a new version. However, I inverted the verbosity of the options of point 1. This is because usually what you want is to affect only your thread. Newer rails versions are capable of running tests in parallel threads, and so you always want to use the thread-aware version during runtime. At initialization time you want the global version so that future threads will inherit the value, but once threads are created you don't want one thread to interfere with another.

File: lib/goldiloader.rb [hunk 4]
```diff
@@ -8,3 +8,37 @@
...
 require 'goldiloader/association_loader'
 require 'goldiloader/active_record_patches'
+
+module Goldiloader
+  @globally_enabled = true
+
+  class << self
+
+    # Sets the process-wide enabled status
+    attr_reader :globally_enabled
```
Comment (jturkel) [inline lib/goldiloader.rb:18]: Should this be an `attr_accessor` to make this writable? 

```suggestion
    attr_accessor :globally_enabled
```

Also the defaulting on line 13 should use this setter (i.e. `self. globally_enabled = true`) to fail fast if we mistype the the variable name.

File: lib/goldiloader.rb [hunk 5]
```diff
@@ -8,3 +8,37 @@
...
+
+    def enabled?
+      Thread.current.fetch(:goldiloader_enabled, @enabled)
```
Comment (jturkel) [inline lib/goldiloader.rb:21]: This isn't defaulting properly to `globally_enabled`. Can we add test coverage for this too?

```suggestion
      Thread.current.fetch(:goldiloader_enabled, globally_enabled)
```

File: lib/goldiloader/active_record_patches.rb [hunk 6]
```diff
@@ -47,7 +47,7 @@ def auto_include!(auto_include = true)
     end
 
     def auto_include_value
-      @values.fetch(:auto_include, true)
+      Goldiloader.enabled? && @values.fetch(:auto_include, true)
```
Comment (jturkel) [inline lib/goldiloader/active_record_patches.rb:50]: Why was this change necessary? Isn't the change to `load_with_auto_include` sufficient?

File: spec/goldiloader/goldiloader_spec.rb [hunk 7]
```diff
@@ -748,6 +748,185 @@
...
+      blog.posts.to_a.first.tags.to_a
+
+      expect(other_blog_post.association(:tags)).not_to be_loaded
+
+      other_blog.posts.each do |post|
+        expect(post.association(:tags)).not_to be_loaded
+      end
+    end
+
+    context "using enabled block" do
```
Comment (jturkel) [inline spec/goldiloader/goldiloader_spec.rb:826]: Can we add a sanity test of the `disabled` block too?

File: spec/goldiloader/goldiloader_spec.rb [hunk 8]
```diff
@@ -748,6 +748,185 @@
...
+            expect(blog.association(:posts_with_inverse_of)).to be_loaded
+            blog.posts_with_inverse_of.each do |post|
+              expect(post.association(:blog)).to be_loaded
+              expect(post.blog.object_id).to eq(blog.object_id)
+            end
+          end
+        end
+      end
+
+      it "marks auto eager loaded models as read only when the association is read only" do
```
Comment (jturkel) [inline spec/goldiloader/goldiloader_spec.rb:914]: This example and the next example seem to duplicate examples [elsewhere](https://github.com/salsify/goldiloader/blob/master/spec/goldiloader/goldiloader_spec.rb#L269) in this file and don't seem particular to this change. Are they necessary?

File: spec/goldiloader/goldiloader_spec.rb [hunk 9]
```diff
@@ -748,6 +748,185 @@
...
+        end
+
+        arr = blog.posts.to_a
+        arr.first.author
+        arr.drop(1).each do |post|
+          expect(post.association(:author)).not_to be_loaded
+        end
+      end
+
+      it "sets inverse associations properly" do
```
Comment (jturkel) [inline spec/goldiloader/goldiloader_spec.rb:899]: Is this testing anything that isn't covered by this [test](https://github.com/salsify/goldiloader/blob/master/spec/goldiloader/goldiloader_spec.rb#L220)? The `Goldiloader.enabled` block seems incidental to what's being tested.

**Review by jturkel — CHANGES_REQUESTED** [review]
The change is looking good. Just a few small things to fix up before it's ready to merge.

File: lib/goldiloader/active_record_patches.rb (hunk 6 above)
Comment (fsateler) [reply lib/goldiloader/active_record_patches.rb:50]: This is a remnant of the previous implementation. With the new change, it doesn't appear to be necessary. Removing.

File: spec/goldiloader/goldiloader_spec.rb (hunk 8 above)
Comment (fsateler) [reply spec/goldiloader/goldiloader_spec.rb:914]: Indeed, this section was born as a copy paste :grimacing:

File: spec/goldiloader/goldiloader_spec.rb [hunk 10]
```diff
@@ -748,6 +748,247 @@
...
+        expect(Goldiloader.enabled?).to be true
+        Goldiloader.enabled = false
+        expect(Goldiloader.enabled?).to be false
+
+        Thread.new { expect(Goldiloader.enabled?).to be true }.join
+      end
+    end
+
+    context "disabled" do
+      after { Goldiloader.globally_enabled = false }
```
[the remark this answers is not in the document]
Comment (fsateler) [reply spec/goldiloader/goldiloader_spec.rb]: Doh! I ran the test manually so I didn't see the failure. The code should be `Goldiloader.globally_enabled = true`. Will set the after block globally to ensure correct status.

```
  after do
    # Always reset enablement status to a known-good state
    Goldiloader.globally_enabled = true
    Goldiloader.enabled = nil
  end
```
